### PR TITLE
GM: Add BSM from Powertrain Bus

### PR DESCRIPTION
**Pull request by garrettpall**
Using fingerprint info, the message is there on most upstream GM car’s except Volts, Escalades ESV, Equinox, and Holden Astra.

Confirmed working on:
Bolt
Silverado
XT4

Prerequisites:
https://github.com/commaai/opendbc/pull/993

### CODE REVIEW

**Comment by ServerDummy** [comment]
GMC 1500(fingerprints as a Silverado) route: dc7716b32bf25574|2023-12-27--21-15-59--0

Bolt EUV route: 3140e1c6195c880e|2023-12-27--21-15-14--0

**Comment by garrettpall** [comment]
On the 2018 Arcadia, the BSM signal is only 1 when the BSM light is on **and the turn signal is on*.

File: selfdrive/car/gm/interface.py [hunk 1]
```diff
@@ -252,6 +253,9 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
       ret.steerActuatorDelay = 0.2
       CarInterfaceBase.configure_torque_tune(candidate, ret.lateralTuning)
 
+    # Detect if BSM message is present
+    ret.enableBsm = BSM_MSG in fingerprint[CanBus.POWERTRAIN]
```
Comment (sshane) [inline selfdrive/car/gm/interface.py:257]:
```suggestion
    ret.enableBsm = 0x142 in fingerprint[CanBus.POWERTRAIN]
```

**Comment by sshane** [comment]
> On the 2018 Arcadia, the BSM signal is only 1 when the BSM light is on **and the turn signal is on**.

Maybe another bit tells us this?

**Comment by garrettpall** [comment]
> Nice find!
> 
> > On the 2018 Arcadia, the BSM signal is only 1 when the BSM light is on **and the turn signal is on**.
> 
> Maybe another bit tells us this?

None in that message, there are times where those are the only bits that change.

**Comment by sshane** [comment]
Have any routes of this problem?

**Comment by garrettpall** [comment]
@sshane df51b2195de66a9b|2023-12-04--15-55-16

**Comment by sshane** [comment]
Looks like the original signals are

```
 SG_ RIghtLaneChangeThreat : 7|1@0+ (1,0) [0|1] ""  XXX
 SG_ LeftLaneChangeThreat : 6|1@0+ (1,0) [0|1] ""  XXX
```

so this makes sense.

There's also some speed signals for the BSMs in there, but no distance.

**Comment by garrettpall** [comment]
@sshane I think I added the note correctly, can you check?

**Comment by sshane** [comment]
lgtm, vivek is trying out the branch on the Bolt for the weekend

**Comment by garrettpall** [comment]
@sshane Everything work out on the Bolt?

**Comment by Verylukyguy** [comment]
> There's also some speed signals for the BSMs in there, but no distance. Can you add a comment to the DBC explaining it can be only when the blinker is on, for some cars?

On the Acadia, the Blind Spot warning light on the mirror lights up when a vehicle is in your blindspot and the warning light flashes if the turn signal is turned on.

**Comment by sshane** [comment]
@Verylukyguy if you want to help us find the bit that has the BSM indicator at all times, you'll need to go get a route where you turn the turn signal on in the direction that it shows the BSM alert for as long as it does. This might confuse some drivers around you though

**Comment by Verylukyguy** [comment]
> @Verylukyguy if you want to help us find the bit that has the BSM indicator at all times, you'll need to go get a route where you turn the turn signal on in the direction that it shows the BSM alert for as long as it does (so we can cross reference it with Cabana). This might confuse some drivers around you though

I definitely want to help improve the overall OpenPilot experience whenever I can.  I'll make a route today and let you guys know when it ready.

**Comment by Verylukyguy** [comment]
Right Side:  df51b2195de66a9b|2024-01-24--17-19-57--8
Left Side:    df51b2195de66a9b|2024-01-24--18-01-06--1

**Comment by sshane** [comment]
I can't find it from that, we'll need at least 10 total seconds throughout a very short route. Remember, whenever the light is on, you need to perform some action, and whenever the light is off, you should stop that action. Maybe put the car in L or N gear so you don't have to use the blinkers?

Also needs to only be for one light per action, so:

left bsm light goes on -> put car in low
left bsm light goes off -> put car back in drive

a few cases for multiple seconds and we should be able to find it

**Comment by Verylukyguy** [comment]
Try this route:  df51b2195de66a9b|2024-01-26--18-54-15--0

**Comment by garrettpall** [comment]
Updated BCMBlindSpotMonitors to BCMBlindSpotMonitor to match merged opendbc code.

**Comment by sshane** [comment]
> Try this route: df51b2195de66a9b|2024-01-26--18-54-15--0

Thanks, unfortunately I still didn't find it. It may only be on another bus on your car. Besides the original signal is Lane Change Threat, so it only being 1 when the turn signal is on doesn't seem crazy.

**Comment by sshane** [comment]
thanks for the PR @garrettpall!

**Pull request merged**